clang/lib/CodeGen/CGDeclCXX.cpp

```diff
@@ -1070,13 +1084,20 @@ CodeGenFunction::GenerateCXXGlobalInitFunc(llvm::Function *Fn,
       // Mark as initialized before initializing anything else. If the
       // initializers use previously-initialized thread_local vars, that's
       // probably supposed to be OK, but the standard doesn't say.
-      Builder.CreateStore(llvm::ConstantInt::get(GuardVal->getType(),1), Guard);
-
-      // The guard variable can't ever change again.
+      // Get the thread-local address via intrinsic.
+      if (IsTLS)
+        GuardAddr = GuardAddr.withPointer(
+            Builder.CreateThreadLocalAddress(Guard.getPointer()),
+            NotKnownNonNull);
+      Builder.CreateStore(llvm::ConstantInt::get(GuardVal->getType(), 1),
+                          GuardAddr);
+
+      // Emit invariant start for TLS guard address.
       EmitInvariantStart(
           Guard.getPointer(),
           CharUnits::fromQuantity(
-              CGM.getDataLayout().getTypeAllocSize(GuardVal->getType())));
+              CGM.getDataLayout().getTypeAllocSize(GuardVal->getType())),
+          IsTLS);
```

Comment:
Just use `GuardAddr` here, please, instead of passing a flag down.

Reply:
Am I allowed to reuse TLS address for invariant start intrinsic (example 1), or I need to access via intrinsic each time (example 2)? If the latter is true, I would need to recalculate the `GuardAddr` again.
example 1:
```
   %tls_addr1 = call align 1 ptr @llvm.threadlocal.address.p0(ptr align 1 @__tls_guard)
   store i8 1, ptr %tls_addr1, align 1
   %3 = call ptr @llvm.invariant.start.p0(i64 1, ptr %tls_addr1)
```
example 2:
```
  %tls_addr1 = call align 1 ptr @llvm.threadlocal.address.p0(ptr align 1 @__tls_guard)
  store i8 1, ptr %tls_addr1, align 1
  %tls_addr2 = call align 1 ptr @llvm.threadlocal.address.p0(ptr align 1 @__tls_guard)
  %4 = call ptr @llvm.invariant.start.p0(i64 1, ptr %tls_addr2)
```

Reply:
Yes, and in fact, sometimes that's required.  The way I think about it is that `threadlocal.address` is a real dynamic operation that resolves the address of the thread-local variable for a specific thread: namely, the current thread at the execution point where `threadlocal.address` appears. When we're doing a complicated operation on a specific thread-local variable, it's important to call `threadlocal.address` at the right point and then use that result throughout the operation.  Otherwise, a suspension in the middle of the operation will leave us working on a different thread-local variable at different points in the operation, which is not semantically allowed.

In this case, we initialize a thread-local variable and then enter an invariant region for it.  We need this to:
1. resolve the address of the TLV for the current thread, prior to performing any initialization;
2. initialize the TLV, which may include suspensions that change the current thread; and
3. enter an invariant region for the TLV we initialized, *not* the TLV for the new thread.  The new thread's TLV may not yet be in an invariant region.

Now, this may actually be moot, because I'm not sure it's actually allowed (or at least I'm not sure it *should* be allowed) to have a coroutine suspension in the middle of the initialization of a thread-local variable.  The interaction of thread-locals with coroutines that actually switch threads is deeply problematic; notably, the TLV for the new thread can actually be uninitialized when you go to use it.  I haven't checked what the standard says here, but honestly this might have to have undefined behavior, in which case we just need to make sure we generate reasonable code as long as the thread *doesn't* change.

Reply:
I had similar concerns. IIUC, the safe(st possible) bet is to generate `threadlocal.address` at each access. In the view of this PR:
1. for the load of `tls_guard`
2. for the store into `tls_guard`
3. for the `invariant.start` intrinsic for `tls_guard`

I am not sure if there is a case where coroutine context switch could happen between `threadlocal.address` and its user instruction. If that is the case, then there is no guarantee for other use cases of `threadlocal.address` as well.
Do you agree or I am missing something?

Reply:
That's exactly backwards — it is semantically incorrect to access different guard variables at different points during the initialization; that does not correctly implement the guard pattern.  Fortunately, I've been able to verify that it doesn't matter, because [expr.await]p3 says:

> An *await-expression* shall not appear in the initializer of a block variable with static or thread storage duration.

Therefore, it is fine to just do a single use of `threadlocal.address`, and it's preferable because it's less code and likely easier to optimize.

(Unfortunately, this restriction doesn't completely eliminate the coherence problems with coroutines and `thread_local`.  In particular, a coroutine that declares a `thread_local` local variable can observe it in an uninitialized state if the thread has changed since the initialization, and I don't see anything in the standard that addresses this.)

Reply:
Ok, thanks for your effort and detailed explanation!
Do you think the best we can do (in this case) is to get `threadlocal.address` from `__tls_guard` at the first access in `__tls_init` function and reuse it in the whole function (example below)?

```
define internal void @__tls_init() {
  %tls_addr = call align 1 ptr @llvm.threadlocal.address.p0(ptr align 1 @__tls_guard)
  %1 = load i8, ptr %1, align 1
  %2 = icmp eq i8 %1, 0
  br i1 %2, label %3, label %5

3:
  store i8 1, ptr %tls_addr, align 1
  %4 = call ptr @llvm.invariant.start.p0(i64 1, ptr %tls_addr)
  call void @__cxx_global_var_init()
  br label %5

5:
  ret void
}
```

Reply:
I feel like we just need to file a DR with WG21 on this general question.  The alternatives I can see:

1. Forbid TLVs in coroutines. Source-breaking.
2. Continue to eagerly initialize local TLVs in coroutines, but bind the name permanently to the TLV that was initialized. This means it is potentially a cross-thread reference after a suspension. It is the user's responsibility to make that not a problem. A semantics change for most (if not at all) existing compilers.
3. Like #2, but give references to the TLV undefined behavior if the thread has changed. The most compatible option, sadly.
4. Like #2, but make the program statically ill-formed if a suspension potentially intervenes between the initialization and a reference to the TLV. Source-breaking. Also a significant new implementation burden for compilers.
5. Switch local TLVs in coroutines to the same lazy-initialization model as global TLVs.  Every reference to the variable dynamically resolves to the TLV for the current thread.  Because of the model change, the variable is always initialized. A semantics change for most (if not all) existing compilers.
6. Like #5, but require the initializer to be a constant expression just to try to minimize the semantic impact.

Note that a cross-thread reference can race not just with read/write or write/write conflicts between threads, but also with the destruction of the other thread.  Because of the latter, it will be a potentially-dangling reference for most async-like use cases.